### DISCUSSION TITLE
More bits transfered for IRremote than expected

### DISCUSSION THREAD

**Discussion by cociweb** [arendst/Tasmota#12781] [Ideas]
Hello,
I'm still trying to configure my proper build for Tasmota with the only the selected protocol what I need.

I'm using the following during the build process based on the suggestion for my previus [question](https://github.com/arendst/Tasmota/discussions/12686):

build_flags             = ${common.build_flags} -DFIRMWARE_IR_CUSTOM -D_IR_ENABLE_DEFAULT_=false -DDECODE_MITSUBISHI112=true -DDECODE_TCL112AC=true -DSEND_TCL112AC=true

(TCL decode uses mitsubishi decode.)
-I hope I did it correctly. Btw, I've never got recognized message with it:

IR_RCV_WHILE_SENDING   is   0 and the IR_RCV_BUFFER_SIZE  is also increased to 512

Anyway, my problem is that the protocol is not recognized, furthermore I'm getting 114 bits insted of the expected 112.
Further info, that the pure library (IRRemoteESP8266 / IRrecvDumpV2.ino) is recognize and decode it without problem (library version: 2.7.19- ).

I've tried to catch the raw data, to examine what is going on. but it seems, something wierd is happening: 
`18:23:54.645 MQT: tele/Tasmota_XXX/RESULT = {"Time":"2021-07-29T18:23:54","IrReceived":{"Protocol":"UNKNOWN","Bits":114,"Hash":"0x84990045","RawData":"+3725-1510+570-1140+545-1110C-455EgEgEdEgE-480EfCfEgCfEgCgE-1135EfE-460+565fCfEhC-425E-1115C-450En+520-1165EgEjCgEgEnC-430CnEhEhEnOhEgCgEgEgEgCgCqEgCqCfEhEgEdEmCgEgEgCgEgEgEdEdEmCfCgEgEgEhEfEjCmEdEnEhEgEg+540jCnEgEgCgEgEgEhEgEgEgCgEgEnEgEhEnEgRjRh+550nEgEjCgEnEgCgEgEgEgCnEgEgEhEfCgEfCgEmCgC","RawDataInfo":[227,227,0]}}`

Anyway, I don't really need the raw data, my main quiestion is what I'm misssed during the compillation, why I am getting 2 more bits only with tasmota? Did I missed something?

Any hint is welcomed.

Thank's.

**Comment by s-hadinger** [comment]
Did you try `tasmota-ir`? It has all protocols enabled.

If the protocol is unknown, the extra bits may be due to signaling bits that are not part of the protocol.

`FIRMWARE_IR_CUSTOM` is no more necessary since #12428

**Comment by Jason2866** [comment]
Please try `tasmota-ir`, if `tasmota-ir` is working, it is "only" your build setup.
We will find out what is missing in that case

**Comment by cociweb** [comment]
Thank you for the suggestions. I've tried with the latest tagged version (9.5.0) but the result not brings the expected result. Unfortunatelly, the prebuilt image gives the same:
```
19:22:39.150 RSL: RESULT = {"IrReceived":{"Protocol":"UNKNOWN","Bits":114,"Hash":"0x546F0A20","Repeat":0}}
19:23:03.713 CMD: SetOption58 1
19:23:03.717 RSL: RESULT = {"SetOption58":"ON"}
19:23:06.440 RSL: RESULT = {"IrReceived":{"Protocol":"UNKNOWN","Bits":114,"Hash":"0x546F0A20","Repeat":0,"RawData":"+3705-1505+570-1140+545-1110C-455E-450+520-480I-1165IjEjEfCfIjCfIjCgE-1115Cf+540-460CfCfI-485CnMlCgEgCdEgEnChEgEgEjEgEgIjCgEgEgEjEgEgEgCgElCgEgIkEgIjMdEdEgEgCgEgEgCgEfEdEdEdMgEgEnCgEgEgCgElCfMdEgEjEgMgEgEjEgEgEjIjEgEgCgEgEgI-505IjIjEgMjEgEgEgCgIjMjIjIjEgMgI-510IjEgEgMdMdIjEnMjMjIjIkE","RawDataInfo":[227,227,0]}}
```

The starting of the message sequence is more or less is okay from the raw timing point of view. 

It is decoded by the library (IRrecvDumpV2.ino) as:
```
Timestamp : 000352.716
Library   : v2.7.19

Protocol  : TCL112AC
Code      : 0x23CB26010024030F380000000083 (112 Bits)
Mesg Desc.: Power: On, Mode: 3 (Cool), Temp: 16C, Fan: 0 (Auto), Econo: Off, Health: Off, Light: On, Turbo: Off, Swing(H): Off, Swing(V): On
uint16_t rawData[227] = {3704, 1510,  566, 1138,  546, 1112,  570, 454,  546, 452,  546, 454,  570, 1112,  544, 456,  570, 454,  546, 1112,  544, 1136,  570, 432,  568, 1112,  544, 480,  546, 452,  546, 1114,  568, 1138,  544, 454,  546, 1114,  568, 1110,  572, 430,  568, 454,  546, 1138,  546, 452,  546, 454,  596, 1112,  544, 456,  572, 454,  546, 452,  546, 454,  570, 432,  568, 452,  546, 452,  546, 456,  570, 432,  568, 454,  546, 452,  544, 458,  570, 454,  544, 454,  544, 454,  568, 456,  546, 452,  546, 1138,  546, 452,  546, 454,  544, 1138,  520, 478,  570, 454,  572, 1114,  570, 1110,  570, 456,  546, 452,  548, 452,  546, 454,  572, 454,  544, 452,  546, 1164,  544, 1138,  546, 1114,  568, 1112,  570, 432,  568, 452,  546, 454,  544, 456,  570, 430,  568, 452,  548, 452,  546, 1138,  546, 1138,  546, 1138,  546, 452,  546, 454,  542, 458,  570, 454,  546, 454,  544, 454,  546, 458,  566, 454,  544, 454,  548, 454,  568, 432,  568, 452,  546, 454,  544, 458,  570, 454,  544, 454,  546, 452,  570, 456,  546, 452,  572, 428,  518, 482,  570, 454,  546, 452,  546, 454,  570, 430,  570, 452,  572, 452,  544, 456,  570, 454,  544, 454,  546, 452,  570, 432,  570, 454,  546, 452,  518, 1164,  546, 1138,  546, 452,  546, 452,  544, 458,  570, 454,  546, 454,  544, 1138,  572};  // TCL112AC
uint8_t state[14] = {0x23, 0xCB, 0x26, 0x01, 0x00, 0x24, 0x03, 0x0F, 0x38, 0x00, 0x00, 0x00, 0x00, 0x83};

```

**Reply by s-hadinger** [reply]
Did you use `tasmota-ir` variant? From your response this is unclear. 

Can you paste here the result of `Status 0`

**Reply by cociweb** [reply]
No, it's not a variant. It's is the tasmota-ir. The standard version was used from [here](https://github.com/arendst/Tasmota/releases/download/v9.5.0/tasmota-ir.bin.gz).

```
18:32:19.428 CMD: Status 0
18:32:19.434 RSL: STATUS = {"Status":{"Module":1,"DeviceName":"Tasmota","FriendlyName":["Tasmota"],"Topic":"tasmota_XXXX","ButtonTopic":"0","Power":0,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0}}
18:32:19.439 RSL: STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota.bin.gz","RestartReason":"Power On","Uptime":"3T23:09:59","StartupUTC":"2021-07-30T18:22:20","Sleep":50,"CfgHolder":4617,"BootCount":6,"BCResetTime":"2021-07-30T19:16:25","SaveCount":14,"SaveAddress":"F6000"}}
18:32:19.443 RSL: STATUS2 = {"StatusFWR":{"Version":"9.5.0(ir)","BuildDateTime":"2021-06-17T08:26:51","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"353/699"}}
18:32:19.446 RSL: STATUS3 = {"StatusLOG":{"SerialLog":2,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["WIFI-AP",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["00008009","2805C8000100060000005A00000000000000","00000180","00006000","00000000"]}}
18:32:19.453 RSL: STATUS4 = {"StatusMEM":{"ProgramSize":597,"Free":404,"Heap":24,"ProgramFlashSize":4096,"FlashSize":4096,"FlashChipId":"16405E","FlashFrequency":40,"FlashMode":3,"Features":["00000809","0F5A0582","04000001","00000002","00000000","00000000","00000000","00000000","00000000"],"Drivers":"1,2,4,5,7,9,10,12","Sensors":""}}
18:32:19.458 RSL: STATUS5 = {"StatusNET":{"Hostname":"tasmota_XXXX-XXXX","IPAddress":"192.168.1.2","Gateway":"192.168.1.1","Subnetmask":"255.255.255.0","DNSServer":"192.168.1.1","Mac":"XX:XX:XX:XX:XX:XX","Webserver":2,"WifiConfig":4,"WifiPower":17.0}}
18:32:19.461 RSL: STATUS6 = {"StatusMQT":{"MqttHost":"","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_9C99E1","MqttUser":"DVES_USER","MqttCount":0,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
18:32:19.466 RSL: STATUS7 = {"StatusTIM":{"UTC":"2021-08-03T17:32:19","Local":"2021-08-03T18:32:19","StartDST":"2021-03-28T02:00:00","EndDST":"2021-10-31T03:00:00","Timezone":"+01:00","Sunrise":"05:26","Sunset":"20:25"}}
18:32:19.469 RSL: STATUS10 = {"StatusSNS":{"Time":"2021-08-03T18:32:19"}}
18:32:19.475 RSL: STATUS11 = {"StatusSTS":{"Time":"2021-08-03T18:32:19","Uptime":"3T23:09:59","UptimeSec":342599,"Vcc":3.504,"Heap":24,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"POWER":"OFF","Wifi":{"AP":1,"SSId":"WIFI-AP","BSSId":"XX:XX:XX:XX:XX","Channel":11,"Mode":"11n","RSSI":100,"Signal":-49,"LinkCount":1,"Downtime":"0T00:00:03"}}}
```

**Comment by s-hadinger** [comment]
Thanks. I don't have any explanation. Tasmota is not only reporting what IRremoteESP8266 finds. Maybe try compiling at 160MHz.

**Reply by cociweb** [reply]
Thank's for the hints. Unfortunatelly, the same wierd thing is happened with 160 MHz.
Could it be a bug? What's your opinion? Should I open an issue?

**Reply by s-hadinger** [reply]
I don't have access to a pc right away. You could try using this https://tasmota.hadinger.fr/until to compare the timings with it dump. See https://tasmota.github.io/docs/IRSend-RAW-Encoding/

**Reply by cociweb** [reply]
Wierd, I've check the tool, what you have suggested, but both capture method gives (more-or-less) the same raw data with the same  227 values (which is odd value). The original library recognize 112bits and the tasmota recognize somehow 114bits from the same raw data. It's worth to mention, that  the IR has 'high/3.3V' value on idle, so the Logical1 is 0V and the logical0 is +3.3V. (but it was exactly the same setup with the pure library as well.)

**Reply by s-hadinger** [reply]
This is good news. What is your `SetOption38` value?

**Reply by cociweb** [reply]
The default: 6 for SetOption38. I've played with it in scope of your suggestion, but gives no changes. Except one fact: if the value is equal or larger than 229 there is no captured data. (nothing). So I revert it to the default 6.

**Reply by s-hadinger** [reply]
Did you try with other protocols?

**Reply by cociweb** [reply]
Let me add to SetOption38 another thought: the received '114' bits are very consecutive, so the hash is always the same for the same buttons. (very good repetitiveness with the default value (6) as well,

As per your suggestion, I've tried a TV IR protocol (32 bits only) with the tasmota-ir. It was OK. (It's worth to mention that TCL112AC is a HVAC protocol with 112 bits.  The HVAC protocol and/or >64bits are handled in diffent way in the library - as I saw. 
[OFF:] Unfortunatelly, I don't have any onther ir remote nor IR led/transmitter. So -as of now - I'm not able to send any custom bits. (But maybe it is not relevant, since the receiveing part is okay. Only the decoding has some glitch with this /type(s) of?/ protocol.)

**Reply by s-hadinger** [reply]
Right. The receiving part seems to be ok. If someone can try with an emitter and a receiver that would be great. I will not be able to try it before 2 weeks
